File: src/main/java/com/temporal/agentic/activities/TemplateNodeActivitiesImpl.java
package com.temporal.agentic.activities;

import com.temporal.agentic.models.*;
import io.temporal.activity.Activity;
import java.util.*;

/**
 * Implementations of template mapping node activities.
 */
public class TemplateNodeActivitiesImpl implements TemplateNodeActivities {

    @Override
    public NodeExecutionResponse executeValidate(NodeExecutionRequest request) {
        String nodeId = request.getNodeId();
        Map<String, Object> input = request.getNodeInput();

        List<String> errors = new ArrayList<>();
        Map<String, Object> outputs = new TreeMap<>();

        String eventType = (String) input.get("eventType");
        String channel = (String) input.get("channel");
        String locale = (String) input.get("locale");

        if (eventType == null || eventType.trim().isEmpty()) {
            errors.add("eventType is required");
        }
        if (channel == null || !List.of("EMAIL", "SMS", "PUSH").contains(channel)) {
            errors.add("channel must be EMAIL, SMS, or PUSH");
        }
        if (locale == null || locale.trim().isEmpty()) {
            errors.add("locale is required");
        }

        outputs.put("isValid", errors.isEmpty());
        outputs.put("validationErrors", errors);

        ContextDelta delta = new ContextDelta();
        delta.setNodeId(nodeId);
        delta.setWrites(new TreeMap<>(input));
        delta.setTimestamp(System.currentTimeMillis());

        NodeExecutionResponse response = new NodeExecutionResponse();
        response.setNodeId(nodeId);
        response.setStatus(errors.isEmpty() ? "SUCCESS" : "FAILED");
        response.setOutputs(outputs);
        response.setContextDelta(delta);
        response.setErrors(errors);
        response.setMetrics(new TreeMap<>());

        Activity.getExecutionContext().heartbeat("VALIDATE completed");
        return response;
    }

    @Override
    public NodeExecutionResponse executeGovernance(NodeExecutionRequest request) {
        String nodeId = request.getNodeId();
        Map<String, Object> input = request.getNodeInput();

        List<String> disclaimers = new ArrayList<>();
        Map<String, Object> outputs = new TreeMap<>();

        String eventType = (String) input.get("eventType");

        if ("FRAUD_ALERT".equalsIgnoreCase(eventType)) {
            disclaimers.add("Reg E Dispute rights notification required");
            disclaimers.add("Consumer account protection disclosure");
        } else if ("STATEMENT_READY".equalsIgnoreCase(eventType)) {
            disclaimers.add("Standard statement availability notice");
        }

        outputs.put("complianceApproved", true);
        outputs.put("disclaimers", disclaimers);

        ContextDelta delta = new ContextDelta();
        delta.setNodeId(nodeId);
        Map<String, Object> writes = new TreeMap<>();
        writes.put("disclaimers", disclaimers);
        delta.setWrites(writes);
        delta.setTimestamp(System.currentTimeMillis());

        NodeExecutionResponse response = new NodeExecutionResponse();
        response.setNodeId(nodeId);
        response.setStatus("SUCCESS");
        response.setOutputs(outputs);
        response.setContextDelta(delta);
        response.setErrors(new ArrayList<>());
        response.setMetrics(new TreeMap<>());

        Activity.getExecutionContext().heartbeat("GOVERNANCE completed");
        return response;
    }

    @Override
    public NodeExecutionResponse executeRetrieve(NodeExecutionRequest request) {
        String nodeId = request.getNodeId();
        Map<String, Object> input = request.getNodeInput();

        Map<String, Object> templates = new TreeMap<>();
        templates.put("STATEMENT_READY", "Your statement for account ending in XXXX is ready.");
        templates.put("FRAUD_ALERT", "We detected suspicious activity on your account.");
        templates.put("ACH_RETURN", "An ACH transfer was returned: {{reason}}");

        Map<String, Object> brandRules = new TreeMap<>();
        brandRules.put("tone", "professional");
        brandRules.put("maxChars", 160);

        Map<String, Object> outputs = new TreeMap<>();
        outputs.put("templates", templates);
        outputs.put("brandRules", brandRules);

        ContextDelta delta = new ContextDelta();
        delta.setNodeId(nodeId);
        delta.setWrites(outputs);
        delta.setTimestamp(System.currentTimeMillis());

        NodeExecutionResponse response = new NodeExecutionResponse();
        response.setNodeId(nodeId);
        response.setStatus("SUCCESS");
        response.setOutputs(outputs);
        response.setContextDelta(delta);
        response.setErrors(new ArrayList<>());
        response.setMetrics(new TreeMap<>());

        Activity.getExecutionContext().heartbeat("RETRIEVE completed");
        return response;
    }

    @Override
    public NodeExecutionResponse executeTemplateMap(NodeExecutionRequest request) {
        String nodeId = request.getNodeId();
        Map<String, Object> input = request.getNodeInput();
        ContextSnapshot contextSnapshot = request.getContextSnapshot();

        String eventType = (String) input.get("eventType");
        Map<String, Object> contextData = contextSnapshot.getData();

        String filledContent;
        if ("FRAUD_ALERT".equalsIgnoreCase(eventType)) {
            filledContent = "We detected suspicious activity on your account. " +
                    "Your recent transaction of $500 at XYZ Store has been flagged.";
        } else {
            filledContent = "Your statement for account ending in 5678 is ready. " +
                    "Please log in to review details.";
        }

        Map<String, Object> selectedTemplate = new TreeMap<>();
        selectedTemplate.put("templateId", eventType);
        selectedTemplate.put("content", filledContent);

        Map<String, Object> outputs = new TreeMap<>();
        outputs.put("selectedTemplate", selectedTemplate);
        outputs.put("filledContent", filledContent);

        ContextDelta delta = new ContextDelta();
        delta.setNodeId(nodeId);
        delta.setWrites(outputs);
        delta.setTimestamp(System.currentTimeMillis());

        NodeExecutionResponse response = new NodeExecutionResponse();
        response.setNodeId(nodeId);
        response.setStatus("SUCCESS");
        response.setOutputs(outputs);
        response.setContextDelta(delta);
        response.setErrors(new ArrayList<>());
        response.setMetrics(new TreeMap<>());

        Activity.getExecutionContext().heartbeat("TEMPLATE_MAP completed");
        return response;
    }

    @Override
    public NodeExecutionResponse executeTransform(NodeExecutionRequest request) {
        String nodeId = request.getNodeId();
        Map<String, Object> input = request.getNodeInput();

        String filledContent = (String) input.get("filledContent");
        String channel = (String) input.get("channel");

        String shortVariant = truncateString(filledContent, 100);
        String longVariant = filledContent;
        String smsLimited = truncateString(filledContent, 160);

        Map<String, Object> outputs = new TreeMap<>();
        outputs.put("shortVariant", shortVariant);
        outputs.put("longVariant", longVariant);
        outputs.put("smsLimited", smsLimited);

        ContextDelta delta = new ContextDelta();
        delta.setNodeId(nodeId);
        delta.setWrites(outputs);
        delta.setTimestamp(System.currentTimeMillis());

        NodeExecutionResponse response = new NodeExecutionResponse();
        response.setNodeId(nodeId);
        response.setStatus("SUCCESS");
        response.setOutputs(outputs);
        response.setContextDelta(delta);
        response.setErrors(new ArrayList<>());
        response.setMetrics(new TreeMap<>());

        Activity.getExecutionContext().heartbeat("TRANSFORM completed");
        return response;
    }

    @Override
    public NodeExecutionResponse executeDecide(NodeExecutionRequest request) {
        String nodeId = request.getNodeId();
        Map<String, Object> input = request.getNodeInput();

        String channel = (String) input.get("channel");
        String selectedVariant;
        String selectionReason;

        if ("SMS".equalsIgnoreCase(channel)) {
            selectedVariant = (String) input.get("smsLimited");
            selectionReason = "SMS requires limited characters";
        } else if ("EMAIL".equalsIgnoreCase(channel)) {
            selectedVariant = (String) input.get("longVariant");
            selectionReason = "Email supports full length content";
        } else {
            selectedVariant = (String) input.get("shortVariant");
            selectionReason = "Default to short variant";
        }

        Map<String, Object> outputs = new TreeMap<>();
        outputs.put("selectedVariant", selectedVariant);
        outputs.put("selectionReason", selectionReason);

        ContextDelta delta = new ContextDelta();
        delta.setNodeId(nodeId);
        delta.setWrites(outputs);
        delta.setTimestamp(System.currentTimeMillis());

        NodeExecutionResponse response = new NodeExecutionResponse();
        response.setNodeId(nodeId);
        response.setStatus("SUCCESS");
        response.setOutputs(outputs);
        response.setContextDelta(delta);
        response.setErrors(new ArrayList<>());
        response.setMetrics(new TreeMap<>());

        Activity.getExecutionContext().heartbeat("DECIDE completed");
        return response;
    }

    @Override
    public NodeExecutionResponse executeAction(NodeExecutionRequest request) {
        String nodeId = request.getNodeId();
        Map<String, Object> input = request.getNodeInput();
        String idempotencyKey = request.getIdempotencyKey();

        String content = (String) input.get("content");
        String channel = (String) input.get("channel");

        // In real scenario, idempotencyKey would be used for DB lookup to prevent duplicates
        String messageId = "MSG-" + System.currentTimeMillis() + "-" + nodeId;

        Map<String, Object> outputs = new TreeMap<>();
        outputs.put("messageId", messageId);
        outputs.put("sendStatus", "SENT");

        ContextDelta delta = new ContextDelta();
        delta.setNodeId(nodeId);
        delta.setWrites(outputs);
        delta.setPromoteKeys(List.of("messageId")); // Promote messageId to top-level
        delta.setTimestamp(System.currentTimeMillis());

        NodeExecutionResponse response = new NodeExecutionResponse();
        response.setNodeId(nodeId);
        response.setStatus("SUCCESS");
        response.setOutputs(outputs);
        response.setContextDelta(delta);
        response.setErrors(new ArrayList<>());
        response.setMetrics(new TreeMap<>());

        Activity.getExecutionContext().heartbeat("ACTION completed: " + messageId);
        return response;
    }

    @Override
    public NodeExecutionResponse executeFinalize(NodeExecutionRequest request) {
        String nodeId = request.getNodeId();
        Map<String, Object> input = request.getNodeInput();
        ContextSnapshot contextSnapshot = request.getContextSnapshot();

        String messageId = (String) input.get("messageId");
        String content = (String) input.get("content");

        Map<String, Object> auditTrail = new TreeMap<>();
        auditTrail.put("messageId", messageId);
        auditTrail.put("timestamp", System.currentTimeMillis());
        auditTrail.put("contentLength", content != null ? content.length() : 0);
        auditTrail.put("contextVersion", contextSnapshot.getContextVersion());

        Map<String, Object> finalResponse = new TreeMap<>();
        finalResponse.put("messageId", messageId);
        finalResponse.put("status", "COMPLETED");

        Map<String, Object> outputs = new TreeMap<>();
        outputs.put("auditTrail", auditTrail);
        outputs.put("finalResponse", finalResponse);

        ContextDelta delta = new ContextDelta();
        delta.setNodeId(nodeId);
        delta.setWrites(outputs);
        delta.setTimestamp(System.currentTimeMillis());

        NodeExecutionResponse response = new NodeExecutionResponse();
        response.setNodeId(nodeId);
        response.setStatus("SUCCESS");
        response.setOutputs(outputs);
        response.setContextDelta(delta);
        response.setErrors(new ArrayList<>());
        response.setMetrics(new TreeMap<>());

        Activity.getExecutionContext().heartbeat("FINALIZE completed");
        return response;
    }

    @Override
    public NodeExecutionResponse executePlanner(NodeExecutionRequest request) {
        String nodeId = request.getNodeId();
        Map<String, Object> input = request.getNodeInput();

        // Mock planner logic: check if SMS channel should trigger transform-sms-limit node
        String channel = (String) input.get("channel");

        Map<String, Object> plan = new TreeMap<>();
        plan.put("version", 1);
        plan.put("planReason", "Initial planning based on channel: " + channel);

        if ("SMS".equalsIgnoreCase(channel)) {
            plan.put("addTransformSmsLimitNode", true);
        }

        Map<String, Object> outputs = new TreeMap<>();
        outputs.put("plan", plan);

        ContextDelta delta = new ContextDelta();
        delta.setNodeId(nodeId);
        delta.setWrites(plan);
        delta.setTimestamp(System.currentTimeMillis());

        NodeExecutionResponse response = new NodeExecutionResponse();
        response.setNodeId(nodeId);
        response.setStatus("SUCCESS");
        response.setOutputs(outputs);
        response.setContextDelta(delta);
        response.setErrors(new ArrayList<>());
        response.setMetrics(new TreeMap<>());

        Activity.getExecutionContext().heartbeat("PLANNER completed");
        return response;
    }

    private String truncateString(String str, int maxLength) {
        if (str == null) return "";
        if (str.length() <= maxLength) return str;
        return str.substring(0, maxLength - 3) + "...";
    }
}

File: src/main/kotlin/com/temporal/agentic/activities/TemplateNodeActivitiesImpl.kt
package com.temporal.agentic.activities

import com.temporal.agentic.models.*
import io.temporal.activity.Activity
import java.util.*

class TemplateNodeActivitiesImpl : TemplateNodeActivities {
    
    override fun executeValidate(request: NodeExecutionRequest): NodeExecutionResponse {
        val nodeId = request.nodeId!!
        val input = request.nodeInput ?: emptyMap()
        
        val errors = mutableListOf<String>()
        val outputs = TreeMap<String, Any>()
        
        val eventType = input["eventType"] as? String
        val channel = input["channel"] as? String
        val locale = input["locale"] as? String
        
        if (eventType.isNullOrBlank()) {
            errors.add("eventType is required")
        }
        if (channel == null || channel !in listOf("EMAIL", "SMS", "PUSH")) {
            errors.add("channel must be EMAIL, SMS, or PUSH")
        }
        if (locale.isNullOrBlank()) {
            errors.add("locale is required")
        }
        
        outputs["isValid"] = errors.isEmpty()
        outputs["validationErrors"] = errors
        
        val delta = ContextDelta(
            nodeId = nodeId,
            writes = TreeMap(input),
            timestamp = System.currentTimeMillis()
        )
        
        Activity.getExecutionContext().heartbeat("VALIDATE completed")
        
        return NodeExecutionResponse(
            nodeId = nodeId,
            status = if (errors.isEmpty()) "SUCCESS" else "FAILED",
            outputs = outputs,
            contextDelta = delta,
            errors = errors,
            metrics = TreeMap()
        )
    }
    
    override fun executeGovernance(request: NodeExecutionRequest): NodeExecutionResponse {
        val nodeId = request.nodeId!!
        val input = request.nodeInput ?: emptyMap()
        
        val disclaimers = mutableListOf<String>()
        val outputs = TreeMap<String, Any>()
        
        val eventType = input["eventType"] as? String
        
        when {
            "FRAUD_ALERT".equals(eventType, ignoreCase = true) -> {
                disclaimers.add("Reg E Dispute rights notification required")
                disclaimers.add("Consumer account protection disclosure")
            }
            "STATEMENT_READY".equals(eventType, ignoreCase = true) -> {
                disclaimers.add("Standard statement availability notice")
            }
        }
        
        outputs["complianceApproved"] = true
        outputs["disclaimers"] = disclaimers
        
        val writes = TreeMap<String, Any>()
        writes["disclaimers"] = disclaimers
        
        val delta = ContextDelta(
            nodeId = nodeId,
            writes = writes,
            timestamp = System.currentTimeMillis()
        )
        
        Activity.getExecutionContext().heartbeat("GOVERNANCE completed")
        
        return NodeExecutionResponse(
            nodeId = nodeId,
            status = "SUCCESS",
            outputs = outputs,
            contextDelta = delta,
            errors = emptyList(),
            metrics = TreeMap()
        )
    }
    
    override fun executeRetrieve(request: NodeExecutionRequest): NodeExecutionResponse {
        val nodeId = request.nodeId!!
        
        val templates = TreeMap<String, Any>()
        templates["STATEMENT_READY"] = "Your statement for account ending in XXXX is ready."
        templates["FRAUD_ALERT"] = "We detected suspicious activity on your account."
        templates["ACH_RETURN"] = "An ACH transfer was returned: {{reason}}"
        
        val brandRules = TreeMap<String, Any>()
        brandRules["tone"] = "professional"
        brandRules["maxChars"] = 160
        
        val outputs = TreeMap<String, Any>()
        outputs["templates"] = templates
        outputs["brandRules"] = brandRules
        
        val delta = ContextDelta(
            nodeId = nodeId,
            writes = outputs,
            timestamp = System.currentTimeMillis()
        )
        
        Activity.getExecutionContext().heartbeat("RETRIEVE completed")
        
        return NodeExecutionResponse(
            nodeId = nodeId,
            status = "SUCCESS",
            outputs = outputs,
            contextDelta = delta,
            errors = emptyList(),
            metrics = TreeMap()
        )
    }
    
    override fun executeTemplateMap(request: NodeExecutionRequest): NodeExecutionResponse {
        val nodeId = request.nodeId!!
        val input = request.nodeInput ?: emptyMap()
        val contextSnapshot = request.contextSnapshot
        
        val eventType = input["eventType"] as? String
        
        val filledContent = when {
            "FRAUD_ALERT".equals(eventType, ignoreCase = true) ->
                "We detected suspicious activity on your account. " +
                "Your recent transaction of \$500 at XYZ Store has been flagged."
            else ->
                "Your statement for account ending in 5678 is ready. " +
                "Please log in to review details."
        }
        
        val selectedTemplate = TreeMap<String, Any>()
        selectedTemplate["templateId"] = eventType ?: ""
        selectedTemplate["content"] = filledContent
        
        val outputs = TreeMap<String, Any>()
        outputs["selectedTemplate"] = selectedTemplate
        outputs["filledContent"] = filledContent
        
        val delta = ContextDelta(
            nodeId = nodeId,
            writes = outputs,
            timestamp = System.currentTimeMillis()
        )
        
        Activity.getExecutionContext().heartbeat("TEMPLATE_MAP completed")
        
        return NodeExecutionResponse(
            nodeId = nodeId,
            status = "SUCCESS",
            outputs = outputs,
            contextDelta = delta,
            errors = emptyList(),
            metrics = TreeMap()
        )
    }
    
    override fun executeTransform(request: NodeExecutionRequest): NodeExecutionResponse {
        val nodeId = request.nodeId!!
        val input = request.nodeInput ?: emptyMap()
        
        val filledContent = input["filledContent"] as? String ?: ""
        
        val shortVariant = filledContent.truncate(100)
        val longVariant = filledContent
        val smsLimited = filledContent.truncate(160)
        
        val outputs = TreeMap<String, Any>()
        outputs["shortVariant"] = shortVariant
        outputs["longVariant"] = longVariant
        outputs["smsLimited"] = smsLimited
        
        val delta = ContextDelta(
            nodeId = nodeId,
            writes = outputs,
            timestamp = System.currentTimeMillis()
        )
        
        Activity.getExecutionContext().heartbeat("TRANSFORM completed")
        
        return NodeExecutionResponse(
            nodeId = nodeId,
            status = "SUCCESS",
            outputs = outputs,
            contextDelta = delta,
            errors = emptyList(),
            metrics = TreeMap()
        )
    }
    
    override fun executeDecide(request: NodeExecutionRequest): NodeExecutionResponse {
        val nodeId = request.nodeId!!
        val input = request.nodeInput ?: emptyMap()
        
        val channel = input["channel"] as? String
        val (selectedVariant, selectionReason) = when {
            "SMS".equals(channel, ignoreCase = true) ->
                (input["smsLimited"] as? String ?: "") to "SMS requires limited characters"
            "EMAIL".equals(channel, ignoreCase = true) ->
                (input["longVariant"] as? String ?: "") to "Email supports full length content"
            else ->
                (input["shortVariant"] as? String ?: "") to "Default to short variant"
        }
        
        val outputs = TreeMap<String, Any>()
        outputs["selectedVariant"] = selectedVariant
        outputs["selectionReason"] = selectionReason
        
        val delta = ContextDelta(
            nodeId = nodeId,
            writes = outputs,
            timestamp = System.currentTimeMillis()
        )
        
        Activity.getExecutionContext().heartbeat("DECIDE completed")
        
        return NodeExecutionResponse(
            nodeId = nodeId,
            status = "SUCCESS",
            outputs = outputs,
            contextDelta = delta,
            errors = emptyList(),
            metrics = TreeMap()
        )
    }
    
    override fun executeAction(request: NodeExecutionRequest): NodeExecutionResponse {
        val nodeId = request.nodeId!!
        val input = request.nodeInput ?: emptyMap()
        val idempotencyKey = request.idempotencyKey
        
        val messageId = "MSG-${System.currentTimeMillis()}-$nodeId"
        
        val outputs = TreeMap<String, Any>()
        outputs["messageId"] = messageId
        outputs["sendStatus"] = "SENT"
        
        val delta = ContextDelta(
            nodeId = nodeId,
            writes = outputs,
            promoteKeys = listOf("messageId"),
            timestamp = System.currentTimeMillis()
        )
        
        Activity.getExecutionContext().heartbeat("ACTION completed: $messageId")
        
        return NodeExecutionResponse(
            nodeId = nodeId,
            status = "SUCCESS",
            outputs = outputs,
            contextDelta = delta,
            errors = emptyList(),
            metrics = TreeMap()
        )
    }
    
    override fun executeFinalize(request: NodeExecutionRequest): NodeExecutionResponse {
        val nodeId = request.nodeId!!
        val input = request.nodeInput ?: emptyMap()
        val contextSnapshot = request.contextSnapshot
        
        val messageId = input["messageId"] as? String
        val content = input["content"] as? String
        
        val auditTrail = TreeMap<String, Any>()
        auditTrail["messageId"] = messageId ?: ""
        auditTrail["timestamp"] = System.currentTimeMillis()
        auditTrail["contentLength"] = content?.length ?: 0
        auditTrail["contextVersion"] = contextSnapshot?.contextVersion ?: 0
        
        val finalResponse = TreeMap<String, Any>()
        finalResponse["messageId"] = messageId ?: ""
        finalResponse["status"] = "COMPLETED"
        
        val outputs = TreeMap<String, Any>()
        outputs["auditTrail"] = auditTrail
        outputs["finalResponse"] = finalResponse
        
        val delta = ContextDelta(
            nodeId = nodeId,
            writes = outputs,
            timestamp = System.currentTimeMillis()
        )
        
        Activity.getExecutionContext().heartbeat("FINALIZE completed")
        
        return NodeExecutionResponse(
            nodeId = nodeId,
            status = "SUCCESS",
            outputs = outputs,
            contextDelta = delta,
            errors = emptyList(),
            metrics = TreeMap()
        )
    }
    
    override fun executePlanner(request: NodeExecutionRequest): NodeExecutionResponse {
        val nodeId = request.nodeId!!
        val input = request.nodeInput ?: emptyMap()
        
        val channel = input["channel"] as? String
        
        val plan = TreeMap<String, Any>()
        plan["version"] = 1
        plan["planReason"] = "Initial planning based on channel: $channel"
        
        if ("SMS".equals(channel, ignoreCase = true)) {
            plan["addTransformSmsLimitNode"] = true
        }
        
        val outputs = TreeMap<String, Any>()
        outputs["plan"] = plan
        
        val delta = ContextDelta(
            nodeId = nodeId,
            writes = plan,
            timestamp = System.currentTimeMillis()
        )
        
        Activity.getExecutionContext().heartbeat("PLANNER completed")
        
        return NodeExecutionResponse(
            nodeId = nodeId,
            status = "SUCCESS",
            outputs = outputs,
            contextDelta = delta,
            errors = emptyList(),
            metrics = TreeMap()
        )
    }
    
    private fun String.truncate(maxLength: Int): String {
        return if (length <= maxLength) this else substring(0, maxLength - 3) + "..."
    }
}
